gpio_controllers/src/gpio_command_controller.cpp

```diff
@@ -0,0 +1,175 @@
+
+#include "hardware_interface/loaned_command_interface.hpp"
+
+namespace gpio_controllers
+{
+using hardware_interface::LoanedCommandInterface;
+
+GpioCommandController::GpioCommandController()
+: controller_interface::ControllerInterface(),
+  rt_command_ptr_(nullptr),
+  gpios_command_subscriber_(nullptr)
+{
+}
+
+CallbackReturn GpioCommandController::on_init()
+{
+  try
+  {
+    auto_declare<std::vector<std::string>>("gpios", std::vector<std::string>());
+    gpio_names_ = node_->get_parameter("gpios").as_string_array();
+    for(std::string &gpio : gpio_names_) 
+      auto_declare<std::vector<std::string>>("command_interfaces." + gpio, std::vector<std::string>());
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+    return CallbackReturn::ERROR;
+  }
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn GpioCommandController::on_configure(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+    gpio_names_ = node_->get_parameter("gpios").as_string_array();
+
+    if (gpio_names_.empty()){
+        RCLCPP_ERROR(get_node()->get_logger(), "'gpios' parameter was empty");
+        return CallbackReturn::ERROR;
+    }
+
+    for(std::string &gpio : gpio_names_){
+      auto interfaces = node_->get_parameter("command_interfaces." + gpio).as_string_array();
+      if (interfaces.empty()){
+        RCLCPP_ERROR(get_node()->get_logger(), "'command_interfaces.%s' parameter was empty",gpio.c_str());
+        return CallbackReturn::ERROR;
+      }
+      if ( !interface_names_.insert( std::make_pair( gpio, interfaces) ).second ) {
+        RCLCPP_ERROR(get_node()->get_logger(), "Trying to override existing gpio setup. Wrong controller parameters.");
+        return CallbackReturn::ERROR;
+      }
+    }
+
+    for (const auto & gpio : gpio_names_){
+      for (const auto & interface_name: interface_names_[gpio]){
+            command_interface_types_.push_back(gpio + "/" + interface_name);
+        }
+    }
+
+    gpios_command_subscriber_ = get_node()->create_subscription<CmdType>(
+        "~/commands", rclcpp::SystemDefaultsQoS(),
+        [this](const CmdType::SharedPtr msg) { rt_command_ptr_.writeFromNonRT(msg); });
+
+    RCLCPP_INFO(get_node()->get_logger(), "configure successful");
+    return CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration
+GpioCommandController::command_interface_configuration() const
+{
+    controller_interface::InterfaceConfiguration command_interfaces_config;
+    command_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+    command_interfaces_config.names = command_interface_types_;
+
+    return command_interfaces_config;
+}
+
+controller_interface::InterfaceConfiguration
+GpioCommandController::state_interface_configuration() const
+{
+  return controller_interface::InterfaceConfiguration{
+    controller_interface::interface_configuration_type::NONE};
```

Comment:
Do we also want to expose `inputs` to the hardware as states (outputs) from the controller?

Reply:
Very good point. For now I am using the `\dynamic_joint_states` topic to do it.

Reply:
To be honest, I am not sure, what would you prefer? @livanov93 and @fmauch, do you have an opinion about this?

Reply:
I am not deep enough (yet) to give a proper answer to this, I'm afraid.

However, having the state of IOs in general sounds like something that should be there. Not only things that are only inputs to the robot, but also the state of outputs (as they might be set by a different place (PLC, Teach pendant, hardware buttons, etc.))

Reply:
OK, I added the feature of publishing the state of commanded gpios as an `control_msgs::msg::InterfaceValue`.